docs/software/ml/pytorch.md

```diff
@@ -0,0 +1,394 @@
+    If this becomes a bottleneck, consider [squashing the venv][ref-guides-storage-venv] into its own memory-mapped, read-only file system to enhance scalability and reduce load times.
+
+Alternatively one can use the uenv as [upstream Spack instance][ref-building-uenv-spack] to to add both Python and non-Python packages.
+However, this workflow is more involved and intended for advanced Spack users.
+
+## Running PyTorch jobs with SLURM
+
+```bash title="slurm sbatch script"
+#!/bin/bash
+#SBATCH --job-name=myjob
+#SBATCH --nodes=1
+#SBATCH --ntasks-per-node=4
+#SBATCH --cpus-per-task=72
+#SBATCH --time=00:30:00
+# (1)!
+#SBATCH --uenv=pytorch/v2.6.0:/user-environment
+#SBATCH --view=default
+
+#################################
+# OpenMP environment variables #
+#################################
+export OMP_NUM_THREADS=8 # (2)!
+
+#################################
+# PyTorch environment variables #
+#################################
+export MASTER_ADDR=$(hostname) # (3)!
+export MASTER_PORT=6000
+export WORLD_SIZE=$SLURM_NPROCS
+export TORCH_NCCL_ASYNC_ERROR_HANDLING=1 # (4)!
+export TRITON_HOME=/dev/shm/ # (5)!
+
+#################################
+# MPICH environment variables   #
+#################################
+export MPICH_GPU_SUPPORT_ENABLED=0 # (6)!
+
+#################################
+# CUDA environment variables    #
+#################################
+export CUDA_CACHE_DISABLE=1 # (7)!
+
+############################################
+# NCCL and Fabric environment variables    #
+############################################
+export NCCL_NET="AWS Libfabric" # (8)!
+export NCCL_NET_GDR_LEVEL=PHB
+export NCCL_CROSS_NIC=1
+export FI_CXI_DISABLE_HOST_REGISTER=1
+export FI_MR_CACHE_MONITOR=userfaultfd
+export FI_CXI_DEFAULT_CQ_SIZE=131072
+export FI_CXI_DEFAULT_TX_SIZE=32768
+export FI_CXI_RX_MATCH_MODE=software
+
+# (9)!
+# (10)!
+srun bash -c "
+    export RANK=\$SLURM_PROCID
+    export LOCAL_RANK=\$SLURM_LOCALID
+    . ./my-venv/bin/activate
+    python myscript.py
+"
+```
+
+1. The `--uenv` option is used to specify the uenv to use for the job.
+   The `--view=default` option is used to load all the packages provided by the uenv.
+2. Set `OMP_NUM_THREADS` if you are using OpenMP in your code.
+   The number of threads should be not greater than the number of cores per task (`$SLURM_CPUS_PER_TASK`).
+   The optimal number depends on the workload and should be determined by testing.
+   Consider for example that typical workloads using PyTorch may fork the processes, so the number of threads should be around the number of cores per task divided by the number of processes.
+3. These variables are used by PyTorch to initialize the distributed backend.
+   The `MASTER_ADDR` and `MASTER_PORT` variables are used to determine the address and port of the master node.
+   Additionally we also need `RANK` and `LOCAL_RANK` but these must be set per-process, see below.
```

Comment:
Actually WORLD_SIZE is needed as well, cf. above and https://pytorch.org/docs/stable/distributed.html#environment-variable-initialization. Maybe a matter of taste, but I always group the rank/world size params together in the SLURM step for better readability.

Reply:
I'll add a comment. The variable is actually set in the sbatch script.